src/Nethermind/Nethermind.Consensus.AuRa/Contracts/TransactionPermissionContract.cs

```diff
@@ -17,112 +17,153 @@
-using System.Linq;
 using Nethermind.Abi;
+using Nethermind.Blockchain.Contracts;
 using Nethermind.Core;
 using Nethermind.Dirichlet.Numerics;
 using Nethermind.Evm;
 
 namespace Nethermind.Consensus.AuRa.Contracts
 {
-    public partial class TransactionPermissionContract : Contract, IActivatedAtBlock
+    public abstract class TransactionPermissionContract : Contract, IVersionedContract
     {
-        private readonly IDictionary<UInt256, ITransactionPermissionVersionedContract> _versionedContracts;
-        private readonly IVersionContract _versionContract;
+        public virtual bool SupportsContractVersion => true;
+
+        public virtual UInt256 ContractVersion(BlockHeader blockHeader)
+        {
+            try
+            {
+                return Constant.Call<UInt256>(blockHeader, nameof(ContractVersion), Address.Zero);
+            }
+            catch (Exception e)
+            {
+                return UInt256.One;;
+            }
+        }
+
+        /// <summary>
+        /// Returns the contract version number needed for node's engine.
+        /// </summary>
+        public abstract UInt256 Version { get; }
+
+        /// <summary>
+        /// Defines the allowed transaction types which may be initiated by the specified sender with
+        /// the specified gas price and data. Used by node's engine each time a transaction is about to be
+        /// included into a block.
+        /// </summary>
+        /// <param name="parentHeader"></param>
+        /// <param name="tx"></param>
+        /// <returns><see cref="TxPermissions"/>Set of allowed transactions types and <see cref="bool"/> If `true` is returned, the same permissions will be applied from the same sender without calling this contract again.</returns>
+        public abstract (TxPermissions Permissions, bool ShouldCache) AllowedTxTypes(BlockHeader parentHeader, Transaction tx);
+
+        protected ConstantContract Constant { get; }
 
         public TransactionPermissionContract(ITransactionProcessor transactionProcessor,
             IAbiEncoder abiEncoder,
             Address contractAddress,
-            long activationBlock,
-            IReadOnlyTransactionProcessorSource readOnlyReadOnlyTransactionProcessorSource) 
+            IReadOnlyTransactionProcessorSource readOnlyReadOnlyTransactionProcessorSource)
             : base(transactionProcessor, abiEncoder, contractAddress)
         {
-            Activation = activationBlock;
-            var constantContract = GetConstant(readOnlyReadOnlyTransactionProcessorSource);
-            _versionedContracts = GetContracts(constantContract).ToDictionary(c => c.Version);
-            _versionContract = (IVersionContract) _versionedContracts.Values.First(c => c is IVersionContract);
+            Constant = GetConstant(readOnlyReadOnlyTransactionProcessorSource);
+        }
+
+        public static TransactionPermissionContractV1 CreateV1(ITransactionProcessor transactionProcessor,
```

Comment:
Move all those static creation methods to VersionedTransactionPermissionContract